Source/Engine/PackageBuilder/PatternLinker.cs

```diff
@@ -130,23 +186,130 @@ protected internal override Syntax VisitEmbeddedPatternReference(EmbeddedPattern
             VisitPatternReference(node);
             return node;
         }
+        
+        protected virtual LinkedPackageSyntax LoadRequiredPackage(string filePath)
+        {
+            if (fPackageCache == null ||
+                !fPackageCache.PackageSyntaxByFilePath.TryGetValue(filePath, out LinkedPackageSyntax linkedPackage))
+            {
+                if (fFileContentProvider == null)
+                    throw PackageLoadError(TextResource.CannotLoadRequiredFile, filePath);
+                string text = fFileContentProvider(filePath);
+                var parser = new SyntaxParser();
+                PackageSyntax package = parser.ParsePackageText(text);
+                linkedPackage = Link(package, Path.GetDirectoryName(filePath), filePath);
+                fPackageCache?.PackageSyntaxByFilePath.Add(filePath, linkedPackage);
+            }
+            return linkedPackage;
+        }
+        
+        protected Exception PackageLoadError(string format, params object[] args)
+        {
+            return new NevodPackageLoadException(string.Format(System.Globalization.CultureInfo.CurrentCulture, 
+                format, args));
+        }
+
+        private bool ValidateRequiredPathAndAddErrors(string filePath, RequiredPackageSyntax requiredPackage)
+        {
+            if (!fRequiredPackageByFilePath.TryAdd(filePath, requiredPackage))
+            {
+                AddError(requiredPackage, TextResource.DuplicatedRequiredPackage, 
+                    requiredPackage.RelativePath, fRequiredPackageByFilePath[filePath].RelativePath);
+                return false;
+            }
+            if (fDependencyStack.Any(x => x != null && x == filePath))
+            {
+                string errorMessage = string.Join(" -> ", fDependencyStack.Reverse()) + " -> " + filePath;
+                AddError(requiredPackage, TextResource.RecursiveFileDependencyIsNotSupported, errorMessage);
+                return false;
+            }
+            return true;
+        }
+        
+        private LinkedPackageSyntax TryLoadRequiredPackage(string filePath, RequiredPackageSyntax node)
+        {
+            LinkedPackageSyntax package = null;
+            try
+            {
+                package = LoadRequiredPackage(filePath);
+            }
+            catch (FileNotFoundException)
+            {
+                AddError(node, TextResource.FileNotFound, filePath);
+            }
+            catch (DirectoryNotFoundException)
+            {
+                AddError(node, TextResource.FileNotFound, filePath);
+            }
+            catch (UnauthorizedAccessException)
+            {
+                AddError(node, TextResource.AccessToFileDenied, filePath);
+            }
+            catch (NevodPackageLoadException)
+            {
+                throw;
+            }
+            catch (Exception)
+            {
+                AddError(node, TextResource.CannotImportFile, filePath);
+            }
```

Comment:
I believe we should not add any `Exception` as error, if it's something unexpected, better to "crash" the program, as it is most likely an internal error (a bug). After `catch (Exception)` removal rethrow may be removed as well, so we map known (expected) exceptions to errors and crash otherwise.